core/src/common/ipc_session.c

```diff
@@ -0,0 +1,241 @@
+ DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+************************************************************************************************************************/
+
+#include "ipc_session.h"
+
+#include <stdlib.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "lwm2m_debug.h"
+#include "lwm2m_list.h"
+#include "lwm2m_util.h"
+
+typedef struct
+{
+    int Sockfd;
+    struct sockaddr FromAddr;
+    int AddrLen;
+
+} IPCChannel;
+
+struct _IPCSession
+{
+    struct ListHead list;
+    IPCSessionID SessionID;
+    IPCChannel RequestChannel;
+    IPCChannel NotifyChannel;
+};
+
+static struct ListHead sessionList;
+
+
+void IPCSession_Init(void)
+{
+    ListInit(&sessionList);
+}
+
+void IPCSession_Shutdown(void)
+{
+    struct ListHead * i, * n;
+    ListForEachSafe(i, n, &sessionList)
+    {
+        IPCSession * session = ListEntry(i, IPCSession, list);
+        if (session != NULL)
+        {
+            free(session);
+        }
+    }
+}
+
+static IPCSession * FindSessionByID(IPCSessionID sessionID)
+{
+    IPCSession * result = NULL;
+    struct ListHead * i;
+    ListForEach(i, &sessionList)
+    {
+        IPCSession * session = ListEntry(i, IPCSession, list);
+        if (session != NULL)
+        {
+            if (session->SessionID == sessionID)
+            {
+                result = session;
+                break;
+            }
+        }
+    }
+    return result;
+}
+
+int IPCSession_New(IPCSessionID sessionID)
+{
+    int result = -1;
+    if (FindSessionByID(sessionID) == NULL)
+    {
+        // add new session record
+        IPCSession * session = malloc(sizeof(*session));
+        if (session != NULL)
+        {
+            memset(session, 0, sizeof(*session));
+            session->SessionID = sessionID;
+            ListAdd(&session->list, &sessionList);
+            result = 0;
+        }
+        else
+        {
+            Lwm2m_Error("Out of memory\n");
+            result = -1;
+        }
+    }
+    else
+    {
+        Lwm2m_Error("Session ID %d is already in use\n", sessionID);
+        result = -1;
+    }
+    return result;
+}
+
+static void AddChannel(IPCChannel * channel, int sockfd, const struct sockaddr * fromAddr, int addrLen)
+{
+    channel->Sockfd = sockfd;
+    channel->FromAddr = *fromAddr;
+    channel->AddrLen = addrLen;
+}
+
+int IPCSession_AddRequestChannel(IPCSessionID sessionID, int sockfd, const struct sockaddr * fromAddr, int addrLen)
+{
+    int result = -1;
```

Comment:
It's a safe default. I prefer not to leave variables undefined at declaration. Although it takes away the compiler's help in finding paths that don't set it, it also acts defensively in case new code is added later. It's mostly a style thing though, and our style has tended to be to specify defaults in most cases. If I change this, there's a lot of other code I'd have to change to match. Ok to leave as-is?